### PR TITLE
New argument: `max_inner_iter`

### PR DESCRIPTION
**Pull request by mlondschien**
contributes to https://github.com/Quantco/glum/issues/850

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -7,9 +7,13 @@
 Changelog
 =========
 
-3.0.3 - unreleased
+3.1.0 - unreleased
 ------------------
 
+**New features:**
```
Comment (lbittarello) [inline CHANGELOG.rst:13]:
```suggestion
**New feature:**
```

File: CHANGELOG.rst [hunk 2]
```diff
@@ -7,9 +7,13 @@
...
+
+- New argument `max_inner_iter` for classes :class:`~glum.GeneralizedLinearRegressor` and :class:`~glum.GeneralizedLinearRegressorCV` to control the maximum number of iterations of the inner solver in the IRLS-CD algorithm.
```
Comment (lbittarello) [inline CHANGELOG.rst:15]:
```suggestion
- New argument ``max_inner_iter`` for classes :class:`~glum.GeneralizedLinearRegressor` and :class:`~glum.GeneralizedLinearRegressorCV` to control the maximum number of iterations of the inner solver in the IRLS-CD algorithm.
```

**Review by lbittarello — APPROVED** [review]

**Review by MarcAntoineSchmidtQC — COMMENTED** [review]
This is great, and was probably originally meant to be exposed. Thanks.

I'm wondering if we should change the default value to 1000 (as mentioned in the linked issue). Can you provide a simulation to show the difference in runtime between the two?

I'm happy to merge this without the change and switch the discussion to another issue.

**Comment by mlondschien** [comment]
> Can you provide a simulation to show the difference in runtime between the two?

I have not managed to reproduce the convergence issues I encountered on a simulated dataset. The dataset I am using is an aggregation of ICU benchmarks. A training and approval by the dataset owner is required to access. So while not proprietary, it's not easy to share.

The defining feature of this dataset is that there are many strongly correlated features and many mean-imputed missing values.

**Comment by lbittarello** [comment]
@stanmart, ping

**Review by stanmart — APPROVED** [review]
LGTM, but maybe don't close #850 until we test if a lower default works better.

**Comment by mlondschien** [comment]
> LGTM, but maybe don't close https://github.com/Quantco/glum/issues/850 until we test if a lower default works better.

My understanding is the the `max_iter_inner` is only needed if there is some problem with the inner solver, e.g., the problem is not strictly convex. Exposing `max_iter_inner` to the user to avoid the solver trying to find a unique minimum that does not exist for many, many iterations is a hotfix. It would be good to solve the underlying problem: Why does the inner solver not converge in say 10 iterations? https://github.com/Quantco/tabmat/pull/408 is a first step. Still, `glum` does not behave as expected for problems such as the one in https://github.com/Quantco/glum/issues/872. Why? Something with the hessian.

**Pull request merged**